test/Microsoft.Azure.WebJobs.Host.UnitTests/Common/BindToStreamTests.cs

```diff
@@ -0,0 +1,461 @@
+                // Convert was never called 
+
+                Assert.Equal("456-123", _log);
+            }
+
+            public Stream Convert(TestStreamAttribute input)
+            {
+                if (input.Access == FileAccess.Read)
+                {
+                    var value = input.Path; // Will exercise the [AutoResolve]
+                    var stream = new MemoryStream(Encoding.UTF8.GetBytes(value));
+                    stream.Position = 0;
+                    return stream;
+                }
+                throw new InvalidOperationException();
+            }
+
+            // Explicit Write access 
+            public void Read(
+                [TestStream(Path="{x}-%y%")] string value
+                )
+            {
+                _log = value;
+            }
+        }
+
+        // If the file does not exist, we should return null from the converter. 
+        // The Framework *can't* assume that an exception translates to NotExist, since there's
+        // no standard exception, so the file might exist but throw a permission denied exception. 
+        [Fact]
+        public void TestNotExist()
+        {
+            TestWorker<ConfigNotExist>();
+        }
+
+        public class ConfigNotExist : IExtensionConfigProvider, ITest<ConfigNotExist>,
+            IConverter<TestStreamAttribute, Stream>
+        {
+            private object _log;
+
+            public void Initialize(ExtensionConfigContext context)
+            {
+                context.AddBindingRule<TestStreamAttribute>().
+                    BindToStream(this, FileAccess.ReadWrite);
+            }
+
+            public void Test(TestJobHost<ConfigNotExist> host)
+            {
+                host.Call("Read1");
+                Assert.Null(_log);
+
+                host.Call("Read2");
+                Assert.Null(_log);
+
+                host.Call("Read3");
+                Assert.Null(_log);
+
+                host.Call("Read4");
+                Assert.Null(_log);
+            }
+
+            public Stream Convert(TestStreamAttribute input)
+            {
+                if (input.Access == FileAccess.Read)
+                {
+                    return null; // Simulate not-exist 
+                }
+                throw new InvalidOperationException();
+            }
+
+            public void Read1([TestStream("path", FileAccess.Read)] Stream value)
+            {
+                _log = value;
+            }
+
+            public void Read2([TestStream] TextReader value)
+            {
+                _log = value;
+            }
+
+            public void Read3([TestStream] string value)
+            {
+                _log = value;
+            }
+
+            public void Read4([TestStream] byte[] value)
+            {
+                _log = value;
+            }
+        }
+
+
+        // Bulk test the success case for different parameter types we can bind to. 
+        [Fact]
+        public void TestStream()
```

Comment:
In general I don't like to have so many test cases covered by one test, but I don't see a clear way to split up the tests without adding far more classes.

Reply:
Exactly :| There's so much machinery to just wire up the extension for testing; it'd be painful to repeat it or every case. 

---
In reply to: [145228086](https://github.com/Azure/azure-webjobs-sdk/pull/1387#discussion_r145228086) [](ancestors = 145228086)